File: app/src/main/java/one/plaza/nightwaveplaza/Api2/ApiClient.java
package one.plaza.nightwaveplaza.Api2;

import okhttp3.FormBody;
import okhttp3.Headers;
import okhttp3.Interceptor;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import one.plaza.nightwaveplaza.Utils.Utils;

public class ApiClient {
    private static final String BASE_URL = "https://api.plaza.one/";

    private final OkHttpClient client;

    public ApiClient() {
        this.client = createClient();
    }

    private OkHttpClient createClient() {
        OkHttpClient.Builder client;
        client = new OkHttpClient.Builder();

        Interceptor headerAuthorizationInterceptor = chain -> {
            okhttp3.Request request = chain.request();
            Headers headers = request.headers().newBuilder().add(
                    "User-Agent", Utils.getUserAgent()).build();
            request = request.newBuilder().headers(headers).build();
            return chain.proceed(request);
        };

        client.addInterceptor(headerAuthorizationInterceptor);
        return client.build();
    }

    public void getStatus(ApiCallback callback) {
        Request request = new Request.Builder().url(BASE_URL + "/status").build();
        this.client.newCall(request).enqueue(callback);
    }

    public void sendReaction(int reaction, String token, ApiCallback callback) {
        FormBody.Builder body = new FormBody.Builder();
        body.add("reaction", String.valueOf(reaction));
        Request request = new Request.Builder()
                .addHeader("Authorization", "Bearer " + token)
                .url(BASE_URL + "/reactions")
                .post(body.build())
                .build();
        this.client.newCall(request).enqueue(callback);
    }
}

File: app/src/main/java/one/plaza/nightwaveplaza/api/ApiClient.kt
package one.plaza.nightwaveplaza.api

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import kotlinx.serialization.json.Json
import okhttp3.OkHttpClient
import okhttp3.Request
import okio.IOException
import one.plaza.nightwaveplaza.BuildConfig
import one.plaza.nightwaveplaza.helpers.Utils
import timber.log.Timber

object ApiClient {
    private val json = Json { ignoreUnknownKeys = true }

    private val baseUrl = BuildConfig.PLAZA_API

    private val client: OkHttpClient by lazy {
        OkHttpClient.Builder()
            .addInterceptor { chain ->
                val request = chain.request().newBuilder()
                    .header("User-Agent", Utils.getUserAgent())
                    .build()
                chain.proceed(request)
            }
            .build()
    }

    private suspend inline fun <reified T> fetch(url: String): T = withContext(Dispatchers.IO) {
        val request = Request.Builder().url(url).build()
        val responseBody: String

        try {
            client.newCall(request).execute().use { response ->
                if (!response.isSuccessful) throw ApiException("HTTP ${response.code}: ${response.message}")
                responseBody = response.body.string()
            }
        } catch (e: IOException) {
            Timber.e(e, "Network execution failed for $url")
            throw ApiException("Network exception", e)
        }

        try {
            json.decodeFromString<T>(responseBody)
        } catch (e: Exception) {
            Timber.e(e, "JSON parse failed for $url")
            throw ApiException("Status parse error", e)
        }
    }

    class ApiException(message: String, cause: Throwable? = null) : Exception(message, cause)

    @Throws(ApiException::class)
    suspend fun getStatus(): Status =
        fetch("$baseUrl/status")

    @Throws(ApiException::class)
    suspend fun getManifest(): UpdateManifest =
        fetch("https://akai.plaza.one/app-view/update-manifest.json")
}